[Miller] 스프링 카페 4단계 - 로그인 구현

rxdcxdrnine opened the pull request:
안녕하세요! 스프링 카페 4단계 PR 을 올린 Miller 입니다 😀😀

### 구현 사항

이번 단계를 구현하면서 추가된 사항들은 다음과 같습니다.

`/main`
- 로그인과 로그아웃 API/뷰를 담당하는 `AuthController` 추가
- 유저 정보 업데이트 API/뷰를 담당하는 `UserController` 의 메서드에 로그인한 유저만 접근 가능하도록 수정
- `DTO` 클래스 네이밍 컨벤션 변경

`/resources`
- `/user/login.html` 파일 추가
- `/components/navbar,html` 에 세션 정보 추가 수정
- `data.sql` 에 초기 데이터 추가

`/test`
- `/unit/controller/AuthControllerTest` 통합 테스트 추가
- `/unit/controller/UserControllerTest` 통합 테스트 시 `MockHttpSession` 활용

### 테스트 결과

![image](https://user-images.githubusercontent.com/50660684/157844438-64a3dab3-1996-4268-9214-ef0e31222ca5.png)

(UserController 통합 테스트의 경우 기존에 작성한 테스트를 수정하고, 세션 관련 테스트를 추가했습니다.)

heroku 배포 링크 :  <https://java-spring-cafe.herokuapp.com>

rxdcxdrnine committed:
feat: controller 에서 session 을 이용한 login, logout 구현

- AuthController 클래스 생성 및 login, logout 메서드 구현
- LoginForm 클래스 생성
- UserService 클래스에서 login 메서드 구현
- User 클래스에서 authenticate 메서드 구현

rxdcxdrnine committed:
feat: /resources/templates 내 login.html 추가 및 navbar.html 내 href 경로 수정

- login.html 파일 추가 및 /components/navbar.html 파일 내 href 경로 수정
- application.properties 내 session 활성화를 위한 설정 추가

rxdcxdrnine committed:
feat: 로그인한 사용자의 유저 정보 수정 기능 추가

- 유저 정 보 수정 시 세션에 저장된 사용자와 일치하는지 확인하는 기능 추가
- 세션에 저장될 유저 정보의 key 값을 SESSION_USER 로 변경

rxdcxdrnine committed:
feat: /resources/data.sql 을 통한 테스트 데이터 추가 및 오류 메시지 수정

rxdcxdrnine committed:
refactor: ~Form 형식의 DTO 이름을 ~Dto 형식으로 변경

rxdcxdrnine committed:
fix: 빌더 패턴에서 필수적인 정보 입력되지 않은 경우 예외 던지기

rxdcxdrnine committed:
refactor: Controller 에서 redirect 시 반환한 ModelAndView 를 제거

rxdcxdrnine committed:
test: Repository 통합 테스트 시 초기 데이터 저장을 제외하기 위한 @Sql 어노테이션 추가

rxdcxdrnine committed:
test: ModelAndView 반환 제거에 따른 Controller 계층 단위/통합 테스트 수정

rxdcxdrnine committed:
test: 회원가입/회원정보 수정 시 session 설정/불러오기에 따른 테스트 수정

- UserControllerTest 단위/통합 테스트에 MockHttpSession 객체 추가
- 테스트 시 마다 session 객체의 SESSION_USER 키에 유저 정보 설정 후 사용

rxdcxdrnine committed:
fix: User 도메인 클래스에서 login 시 DTO 에 의존하지 않도록 변경

- User 및 UserService 클래스의 authenticate 메서드 파라미터 변경

rxdcxdrnine committed:
test: AuthControllerTest 추가 및 UserControllerTest 내 유저 정보 업데이트 테스트 추가

- AuthControllerTest 내 로그인 폼 요청, 로그인, 로그아웃 테스트 추가
- UserControllerTest 내 유저 정보 업데이트 폼 요청, 유저 정보 업데이트 테스트 추가
- 위 Controller 계층의 단위/통합 테스트 진행
- unit/UserControllerTest 에서 UserSetUp 클래스를 외부 클래스로 분리
- mockMvc 로 요청 시 테스트와 관련없는 필요없는 파라미터 제거

rxdcxdrnine committed:
refactor: google style guide 적용

rxdcxdrnine committed:
fix: EOF new line 추가

rxdcxdrnine committed:
feat: data.sql 내 article 테이블 초기 데이터 추가

rxdcxdrnine committed:
fix: Controller 통합 테스트 시 data.sql 초기 데이터 삽입을 적용하지 않기 위한 @Sql 어노테이션 추가

rxdcxdrnine committed:
test: AuthControllerTest 단위/통합테스트에 @DisplayName 어노테이션 추가

rxdcxdrnine committed:
test: 서비스 계층 login 메서드 테스트 추가

- /unit/service/mock/UserServiceTest.java 내 mockito 의 mock 을 이용한 login 테스트 추가
- /unit/service/stub/UserServiceTest.java 내 stub 을 이용한 login 테스트 추가

rxdcxdrnine committed:
test: build.gradle 내 jacoco 추가 및 report, verification 설정 등록

rxdcxdrnine committed:
docs: README.md 내 API 명세 추가

rxdcxdrnine committed:
docs: jacoco 리포트 이미지 추가

wheejuni commented on src/main/java/com/kakao/cafe/controller/UserController.java:
```diff
@@ -43,29 +49,38 @@ public String formCreateUser() {
-        User user = userService.updateUser(userForm);
+    public String updateUser(@PathVariable String userId, UserDto userDto,
+        HttpSession session) {
+        confirmSession(session, userId);
+
+        userDto.setUserId(userId);
+        userService.updateUser(userDto);
```
우선 유저의 요청으로 넘어온 DTO 클래스에 컨트롤러에서 인위적으로 값 변경을 하는 방향이 좋지는 않은 것 같습니다.
DTO는 불변함을 유지하면서 가급적 세터 호출을 안하는 방향으로 갔으면 좋겠고요..
서비스에 세션 저장된 `User` 까지 같이 넘겨서, 서비스에서 유효성 체크를 하고 업데이트 로직을 수행해야 하지 않을까 의견 드립니다.

rxdcxdrnine replied:
감사합니다! 말씀하신 대로 서비스에 세션 저장된 `User` 를 서비스 계층으로 넘기고, 서비스 계층에서 비밀번호 유효성 검증을 하고 업데이트 로직을 수행하도록 수정했습니다.

wheejuni commented on src/main/java/com/kakao/cafe/domain/Article.java:
```diff
@@ -48,6 +50,10 @@ public Builder createdDate(LocalDateTime createdDate) {
         }
 
         public Article build() {
+            if (writer == null || title == null | contents == null) {
+                throw new NotFoundException(ErrorCode.FIELD_NOT_FOUND);
+            }
```
`build` 라는 메서드명도 모호하긴 한데 그 모호한 메서드명과도 부합이 안되는 로직이 담겨있군요.

rxdcxdrnine replied:
`Article` 클래스에서 `Builder` 패턴으로 `build` 시에, 필수적으로 입력해야 하는 인자가 입력되지 않은 경우에 런타임에 `FIELD_NOT_FOUND` 에러가 발생하도록 했습니다. 하지만 생각해보니 컴파일 시점이 아닌 런타임 시점에 이 부분을 검사하고 에러가 발생하는 것 자체가 잘못된 것 같습니다. `Builder` 클래스 내에서 필수적으로 입력해야 하는 인자는 `Builder` 의 생성자에서 받고, 선택적으로 입력하는 인자는 `setter` 메서드에서 받도록 수정하겠습니다.

rxdcxdrnine replied:
빌더 클래스에서 선택적으로 입력하는 인자보다 필수적으로 입력하는 인자가 많아, 빌더 패턴을 제거하고 점층적 생성자 패턴을 적용했습니다!

wheejuni commented on src/main/java/com/kakao/cafe/domain/User.java:
```diff
@@ -97,6 +102,13 @@ public User update(User user) {
     }
 
+    public User authenticate(String userId, String password) {
+        if (!this.userId.equals(userId) || !this.password.equals(password)) {
+            throw new InvalidRequestException(ErrorCode.INCORRECT_USER);
+        }
```
`userId` 를 기준으로 이 객체를 탐색해서 여기 메소드에 진입하지 않을까요?
`userId` 를 또 검증하는 이유가 있나요?

rxdcxdrnine replied:
정확한 지적 감사합니다! 이미 `userId` 는 일치하기 때문에, `password` 에 대해서만 검증하도록 하겠습니다.

wheejuni commented on src/main/java/com/kakao/cafe/dto/ArticleDto.java:
```diff
@@ -1,6 +1,6 @@
 package com.kakao.cafe.dto;
 
-public class ArticleForm {
+public class ArticleDto {
```
`Dto` 는 사실 좀 애매하긴 합니다. 무엇을 위한 DTO인지... 글을 읽기 위한 건지, 게시하기 위한 건지 잘 알 수 없고요.
가급적이면 `Request` 혹은 `Response` 라는 접미사를 사용해서 요청측인지 응답측인지 정확히 구분하는 것도 좋고요.
아무튼 클래스명이 조금 더 정확했으면 합니다.

rxdcxdrnine replied:
정확한 지적 감사합니다! DTO 클래스 명을 `ArticleSaveRequest` 와 같이 목적과 요청/응답을 구분할 수 있도록 수정하겠습니다.

wheejuni commented on src/main/java/com/kakao/cafe/service/ArticleService.java:
```diff
@@ -21,12 +21,12 @@ public ArticleService(ArticleRepository articleRepository, UserRepository userRe
     }
 
-    public Article write(ArticleForm articleForm) {
+    public Article write(ArticleDto articleDto) {
```
`Article` 은 도메인 객체인데 리턴값이 도메인 객체로 나가면 받아쓰는 쪽에서는 자기 계층을 넘어선 클래스에 의존성을 갖게 되긴 하겠네요.
어려운 문제긴 한데.... 음....

rxdcxdrnine replied:
'자기 계층을 넘어선 클래스에 의존성을 갖는다' 는 예를 들면 `UserController` 계층에서 `ArticleService` 의 `write()` 을 호출할 때, `Article` 객체를 받기 때문에 `Controller` 계층을 넘어서 `Article` 도메인 클래스에 의존성을 가진다 라고 생각하면 될까요?  

`Service` 계층에서 도메인 객체가 아닌 DTO 를 반환해 계층 간의 의존성을 느슨하게 만들 수 있는 장점이 있을 것이라고 생각이 드는데, 처음에는 DTO 가 아닌 도메인 객체를 반환해야 `Service` 계층을 사용하는 여러 `Controller` 계층에서 각자 필요한 정보를 모두 통일성 있게 제공할 수 있을 것이라고 생각했습니다.

2가지의 장점을 모두 가질 수 있도록, 도메인 객체의 모든 필드를 포함한 DTO 를 반환하도록 수정해보겠습니다. `Controller` 계층과 `Domain` 계층 간의 의존성을 한 번도 생각해본 적이 없는데, 주제를 던져주셔서 감사합니다!

wheejuni commented on src/main/java/com/kakao/cafe/service/UserService.java:
```diff
@@ -19,9 +20,9 @@ public UserService(UserRepository userRepository) {
-        return userRepository.save(Mapper.map(userForm, User.class));
+    public User register(UserDto userDto) {
+        validateUserId(userDto.getUserId());
+        return userRepository.save(Mapper.map(userDto, User.class));
```
`repository` 로 도메인 클래스의 저장 책임을 격리시켰는데 `Mapper` 는 왜 등장하는지 모르겠네요?
만약 매퍼의 작동이 필요하다면 레포지토리 안에서 이뤄져야하는거 아닐지 모르겠습니다. `UserRepository` 라는 `User` 클래스의 저장만을 담당해야 할 클래스가 `Mapper` 를 받고, 심지어 클래스 패턴 매칭(?) 까지 하는 듯한 모습이 굉장히 굉장히 어색해 보입니다...

rxdcxdrnine replied:
정확한 지적 감사합니다! `register()` 메서드의 인자로 `Mapper` 클래스를 이용해 입력된 DTO 를 엔티티로 변환하고, `UserRepository` 의 `save()` 메서드에서는 변환된 엔티티를 인자로 입력받도록 했습니다. 하지만 말씀하신 것처럼 변환 로직이 `UserRepository` 까지 전달되는 것처럼 보이므로, 변환 로직은 서비스 메서드의 상위에서 처리하고, 아래에는 변환된 엔티티를 처리하고 리포지토리 계층의 메서드로 전달하도록 구성하겠습니다.

wheejuni reviewed (changes requested):
수고 많으셨습니다. 💯 
제가 앞선 단계 진행하신 걸 리뷰하지 않아서 맥락에 맞지 않는 코멘트를 남겼을 수도 있습니다.
궁금한 부분도 있고 개선해야 할 점들도 있는 것 같아 코멘트 남겼으니 확인 부탁드립니다~

rxdcxdrnine committed:
fix: DTO 계층 이름 변경 및 Service 계층 내 Response 객체 반환 변경

- ArticleDto, UserDto, LoginDto 를 각각 ArticleSaveRequest, UserSaveRequest, UserLoginRequest 로 변경
- Service 계층의 모든 메서드에서 Response 객체를 반환하도록 변경 및 주석 추가
- Response 반환 후 비교를 위해 equals 메서드 오버라이딩
- 위의 사항에 따라 테스트 코드를 수정

rxdcxdrnine committed:
fix: Article, User 객체 내 빌더 재거 및 점층적 생성자 구현

- 테스트 코드 내 내 빌더 제거 및 생성자로 대체

rxdcxdrnine committed:
fix: UserService 클래스 내 updateUser 시 request, user 객체 인자를 받도록 수정

- updateUser 내에서 user 객체 내부에서 request 의 password 검증, 데이터베이스에서 user 객체를 찾아 요청된 필드를 수정 후 저장
- User 객체 내 update 메서드 내에서 검증 메서드 제거 및 제거로 인한 일부 테스트 삭제
- updateUser 객체 변경에 따른 테스트 코드 수정

rxdcxdrnine committed:
fix: session User 반환 시 User 객체가 아닌 UserResponse 객체 반환하도록 수정

- updateUser 에서 UserResponse 객체를 인자로 입력하도록 수정
- updateUser 함수 변경에 따른 테스트 코드 수정

rxdcxdrnine commented:
리뷰 감사합니다! @wheejuni 

지적해주신 부분에 따라, Builder 클래스와 DTO 클래스들을 수정하고 해당 클래스를 사용하는 테스트 코드도 전부 수정했습니다.
re-request 를 요청 드렸으니 확인 부탁드립니다 😀😀
